packages/editor/src/editor/browser/extract/ExtractComponentDialog.tsx

```diff
@@ -0,0 +1,159 @@
+  openDialog: boolean;
+  setOpenDialog: Dispatch<SetStateAction<boolean>>;
+};
+
+export const ExtractComponentDialog = ({ children, data, openDialog, setOpenDialog }: ExtractComponentDialogProps) => {
+  const { t } = useTranslation();
+  const layoutId = (data.config as Layout)?.id?.length > 0 ? (data.config as Layout).id : data.cid;
+
+  return (
+    <Dialog open={openDialog} onOpenChange={setOpenDialog}>
+      <DialogTrigger asChild>{children}</DialogTrigger>
+      <DialogContent onClick={e => e.stopPropagation()}>
+        <DialogHeader>
+          <DialogTitle>{t('dialog.extractComponent', { component: layoutId })}</DialogTitle>
+        </DialogHeader>
+        <ExtractComponentDialogContent data={data} layoutId={layoutId} />
+      </DialogContent>
+    </Dialog>
+  );
+};
+
+const ExtractComponentDialogContent = ({ data, layoutId }: { data: Component | ComponentData; layoutId: string }) => {
+  const { context, namespace } = useAppContext();
+  const { t } = useTranslation();
+  const queryClient = useQueryClient();
+  const { validateComponentName, validateComponentNamespace } = useExtractFieldValidation();
+  const variableInfo = useMeta('meta/data/attributes', context, { types: {}, variables: [] }).data;
+
+  const [name, setName] = useState(layoutId);
+  const [nameSpace, setNameSpace] = useState(namespace);
+  const [field, setField] = useState('data');
+
+  const [tree, setTree] = useState<Array<BrowserNode<Variable>>>([]);
+  const nameValidation = useMemo(() => validateComponentName(name), [name, validateComponentName]);
+  const namespaceValidation = useMemo(() => validateComponentNamespace(nameSpace), [nameSpace, validateComponentNamespace]);
+  const buttonDisabled = useMemo(
+    () => nameValidation?.variant === 'error' || namespaceValidation?.variant === 'error',
+    [nameValidation, namespaceValidation]
+  );
+
+  const extractIntoComponent = useFunction(
+    'extractIntoComponent',
+    {
+      context,
+      layoutId: data.cid,
+      newComponentName: name,
+      nameSpace: nameSpace,
+      dataClassField: field
+    },
+    {
+      onSuccess: componentName => {
+        toast.info('Ivy Component "' + componentName + '" was successfully created');
+        queryClient.invalidateQueries({ queryKey: genQueryKey('data', context) });
+      },
+      onError: error => {
+        toast.error('Failed to extract Component into own Ivy Component', {
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          description: (error as any).data?.includes('already exists')
+            ? 'Ivy Component ' + ((data.config as Layout).id.length > 0 ? (data.config as Layout).id : data.cid) + ' already exists'
+            : error.message
+        });
+      }
+    }
+  );
+
+  useEffect(() => {
+    const fullTree = variableTreeData().of(variableInfo);
+    setTree(fullTree[0]?.children && fullTree[0].children.length > 0 ? [fullTree[0], ...fullTree[0].children] : fullTree);
+  }, [variableInfo]);
```

Comment:
you only set the tree here. so instead of a useState + useEffect you can easily use a useMemo hook.

Reply:
Also is it not realla a tree here but a flat list, so I think you should better name this variable what it is. e.g. dataClassItems or something like that